File: packages/eyes-universal-poc/src/socket.js
const WebSocket = require('ws')
const uuid = require('uuid')
const chalk = require('chalk')

function makeSocket(ws) {
  let socket = null
  const listeners = new Map()
  const queue = new Set()

  attach(ws)

  function attach(ws) {
    if (!ws) return

    if (ws.readyState === WebSocket.CONNECTING) ws.on('open', () => attach(ws))
    else if (ws.readyState === WebSocket.OPEN) {
      socket = ws
      queue.forEach(command => command())
      queue.clear()

      socket.on('message', message => {
        const {name, key, payload} = deserialize(message)
        const fns = listeners.get(name)
        if (fns) fns.forEach(fn => fn(payload, key))
        if (key) {
          const fns = listeners.get(`${name}/${key}`)
          if (fns) fns.forEach(fn => fn(payload, key))
        }
      })
      socket.on('close', () => {
        const fns = listeners.get('close')
        if (fns) fns.forEach(fn => fn())
      })
    }
  }

  function connect(url) {
    const ws = new WebSocket(url)
    attach(ws)
  }

  function disconnect() {
    if (!socket) return
    socket.terminate()
    socket = null
  }

  function request(name, payload) {
    return new Promise((resolve, reject) => {
      const key = uuid.v4()
      console.log(`${chalk.blue('[REQUEST]')} ${name}, ${key}, ${JSON.stringify(payload, null, 2)}`)
      emit({name, key}, payload)
      once({name, key}, response => {
        if (response.error) return reject(response.error)
        return resolve(response.result)
      })
    })
  }

  function command(name, fn) {
    on(name, async (payload, key) => {
      try {
        console.log(
          `${chalk.yellow('[COMMAND]')} ${name}, ${key}, ${JSON.stringify(payload, null, 2)}`,
        )
        const result = await fn(payload)
        emit({name, key}, {result})
      } catch (error) {
        emit({name, key}, {error})
      }
    })
  }

  function subscribe(name, publisher, fn) {
    const subscription = uuid.v4()
    emit(name, {publisher, subscription})
    const off = on({name, key: subscription}, fn)
    return () => (emit({name, key: subscription}), off())
  }

  function on(type, fn) {
    const name = typeof type === 'string' ? type : `${type.name}/${type.key}`
    let fns = listeners.get(name)
    if (!fns) {
      fns = new Set()
      listeners.set(name, fns)
    }
    fns.add(fn)
    return () => off(name, fn)
  }

  function once(type, fn) {
    const off = on(type, (...args) => (fn(...args), off()))
    return off
  }

  function off(name, fn) {
    if (!fn) return listeners.delete(name)
    const fns = listeners.get(name)
    if (!fns) return false
    const existed = fns.delete(fn)
    if (!fns.size) listeners.delete(name)
    return existed
  }

  function emit(type, payload) {
    const command = () => socket.send(serialize(type, payload))
    if (socket) command()
    else queue.add(command)
    return () => queue.delete(command)
  }

  function ref() {
    const command = () => socket._socket.ref()
    if (socket) command()
    else queue.add(command)
    return () => queue.delete(command)
  }

  function unref() {
    const command = () => socket._socket.unref()
    if (socket) command()
    else queue.add(command)
    return () => queue.delete(command)
  }

  return {
    on,
    once,
    off,
    emit,
    request,
    command,
    subscribe,
    connect,
    disconnect,
    ref,
    unref,
  }
}

function serialize(type, payload) {
  const message =
    typeof type === 'string' ? {name: type, payload} : {name: type.name, key: type.key, payload}
  return JSON.stringify(message)
}

function deserialize(message) {
  return JSON.parse(message)
}

module.exports = makeSocket

File: rb/eyes-selenium/lib/socket.rb
require 'faye/websocket'
require 'json'

module Applitools
  class Socket
    attr_reader :listeners, :queue

    def initialize
      @listeners = {}
      @queue = []
    end

    def connect(uri, ws)
      @socket = ws ? ws : ::Faye::WebSocket::Client.new(uri)

      queue.each {|command| command.call}
      queue.clear

      ws.on :message do |event|
        message = JSON.parse(event.data).values
        params = [message[:payload], message[:key]]
        find_and_execute_listeners_by_name(message[:name], params)
        find_and_execute_listeners_by_name("#{name}/#{key}", params) if (key)
      end

      ws.on :close do |event|
        find_and_execute_listeners_by_name('close')
      end
    end

    def emit(message, payload)
      command = ->() {@socket.send(serialize(message, payload))}
      @socket ? command.call : queue.push(command)
    end

    def command(name, fn)
      on(name, ->(payload, key) {
        begin
          puts "[COMMAND] #{name}, #{key}, #{JSON.generate(payload)}"
          result = fn.call(payload)
          emit({name: name, key: key}, result)
        rescue => error
          emit({name: name, key: key}, error.message)
        end
      })
    end

    private

      def find_and_execute_listeners_by_name(name, params = [])
        fns = listeners[name.to_sym]
        return if (!fns)
        fns.each {|fn| fn.call(*params)}
      end

      def serialize(type, payload)
        message = type.is_a?(String) ? 
          {:name => type, :payload => payload} : {:name => type[:name], key: type[:key], :payload => payload}
        JSON.generate(message)
      end

      def on(type, fn)
        name = type.is_a?(String) ? type : "#{type[:name]}/#{type[:key]}"
        fns = listeners[name.to_sym]
        if (!fns)
          fns = []
          listeners[name.to_sym] = fns
        end
        fns.push(fn)
      end

  end
end
